### PR TITLE
[TS] Better shard balancing

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
The existing `Cell` hashCode is always 0 if rowName and colName match. This could be the cause of write->shard assignment imbalances that we've seen in targeted sweep. This should make the distribution more uniform.

**Implementation Description (bullets)**:
Implements hashCode for `CellReference` as if the `Cell` hashCode was generated by immutables, without changing the `Cell.hashCode()` impl.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Some existing tests needed to be changed as they used the hashCode hack. No other tests as this is a super standard implementation, but existing tests don't break

**Concerns (what feedback would you like?)**:
None really, I was considering having a separate method, but this has the added benefit of also working better if we have a `CellReference` keyed HashMap so decided to go with this.

**Where should we start reviewing?**:
small

**Priority (whenever / two weeks / yesterday)**:
soon^{tm}

### CODE REVIEW

**Review by jeremyk-91 — APPROVED** [review]
🕐 0:16 🏁 1:20

Eighth PR of the day. In any case this one is close - some comments on changelog/testing but we're basically there.

Approving but removed MWR - please fix the needed bits.

File: changelog/@unreleased/pr-5274.v2.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+type: improvement
+improvement:
+  description: Improved `CellReference.hashCode()` method to use a more standard implementation rather than depend on `Cell.hashCode()`. This fixes imbalance in the targeted sweep queue caused by writes to cells where row and column names match.
```
[the remark this answers is not in the document]
Comment (gmaretic) [reply changelog/@unreleased/pr-5274.v2.yml:3]: The impl that immutables uses was in effective java, so I consider it pretty standard I guess

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/AbstractSweepQueueTest.java [hunk 2]
```diff
@@ -150,16 +155,24 @@ void putTimestampIntoTransactionTable(long ts, long commitTs) {
...
+    static CellReference getCellRefWithFixedShard(long seed, TableReference tableRef, int shards) {
+        byte[] rowName = PtBytes.toBytes(seed);
+        int counter = 0;
+        CellReference cellReference;
+        do {
+            byte[] colName = PtBytes.toBytes(counter);
+            cellReference = CellReference.of(tableRef, Cell.create(rowName, colName));
+            counter++;
+        } while (IntMath.mod(cellReference.hashCode(), shards) != 0);
+        return cellReference;
```
[the remark this answers is not in the document]
Comment (gmaretic) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/AbstractSweepQueueTest.java:175]: nice!

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/AbstractSweepQueueTest.java [hunk 3]
```diff
@@ -150,16 +155,24 @@ void putTimestampIntoTransactionTable(long ts, long commitTs) {
...
-        return Cell.create(PtBytes.toBytes(seed), PtBytes.toBytes(seed));
+    static CellReference getCellRefWithFixedShard(long seed, TableReference tableRef, int shards) {
+        byte[] rowName = PtBytes.toBytes(seed);
+        int counter = 0;
+        CellReference cellReference;
+        do {
+            byte[] colName = PtBytes.toBytes(counter);
+            cellReference = CellReference.of(tableRef, Cell.create(rowName, colName));
+            counter++;
+        } while (IntMath.mod(cellReference.hashCode(), shards) != 0);
```
Comment (berler) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/AbstractSweepQueueTest.java:174]: Is it a problem that this function as written becomes more expensive as the number of shards increases?

**Pull request merged**